contract_comm/evm.go

```diff
@@ -124,6 +130,64 @@ func Transfer(db vm.StateDB, sender, recipient common.Address, amount *big.Int)
 	db.AddBalance(recipient, amount)
 }
 
+// GetParentSealFn returns a GetParentSeal function that returns the aggregated parent seal at the given block number.
+// Note: Unless previously cached, retrieves every block between the reference block and n.
+func GetParentSealBitmapFn(ref *types.Header, chain ChainContext) func(uint64) *big.Int {
+	var cache map[uint64]*big.Int
+
+	return func(n uint64) *big.Int {
+		// If the block is the unsealed reference block or later, return nil.
+		if n >= ref.Number.Uint64() {
+			return nil
+		}
+
+		// If there's no cache yet, make one
+		if cache == nil {
+			cache = make(map[uint64]*big.Int)
+		} else {
+			// Try to fulfill the request from the cache
+			if bitmap, ok := cache[n]; ok {
+				return bitmap
+			}
+		}
+
+		// Not cached, iterate the blocks and cache the hashes (not limited here)
+		for header := chain.GetHeader(ref.ParentHash, ref.Number.Uint64()-1); header != nil; header = chain.GetHeader(header.ParentHash, header.Number.Uint64()-1) {
+			var bitmap *big.Int
+			istanbulExtra, err := types.ExtractIstanbulExtra(header)
+			if err == nil {
+				bitmap = istanbulExtra.AggregatedSeal.Bitmap
+				cache[header.Number.Uint64()] = bitmap
+			}
+			if n == header.Number.Uint64() {
+				return bitmap
+			}
+		}
+		return nil
```

Comment:
Should we ever get here?

Reply:
Only if something goes very wrong.

Reply:
Should we log and return an error instead?